test/template/test_template.py

```diff
@@ -153,21 +154,66 @@ def test_template_nominal_csv(tmp_path):
     tmpl = Template(format="A4", title="Sample Invoice")
     tmpl.parse_csv(HERE / "mycsvfile.csv", delimiter=";")
     tmpl.add_page()
+    tmpl["empty_fields"] = "empty"
     assert_pdf_equal(tmpl, HERE / "template_nominal_csv.pdf", tmp_path)
     tmpl = Template(format="A4", title="Sample Invoice")
     tmpl.parse_csv(HERE / "mycsvfile.csv", delimiter=";", encoding="utf-8")
     tmpl.add_page()
+    tmpl["empty_fields"] = "empty"
     assert_pdf_equal(tmpl, HERE / "template_nominal_csv.pdf", tmp_path)
 
 
+def test_template_multipage(tmp_path):
+    """Testing a Template() populating several pages."""
+    tmpl = Template(format="A4", title="Sample Invoice")
+    tmpl.parse_csv(HERE / "mycsvfile.csv", delimiter=";")
+    tmpl.add_page()
+    tmpl["name0"] = "Joe Doe"
+    tmpl["title0"] = "Director"
+    tmpl.add_page()
+    tmpl["name0"] = "Jane Doe"
+    tmpl["title0"] = "General Manager"
+    tmpl.add_page()
+    tmpl["name0"] = "Heinz Mustermann"
+    tmpl["title0"] = "Worker"
+    assert_pdf_equal(tmpl, HERE / "template_multipage.pdf", tmp_path)
+
+
+# pylint: disable=unused-argument
+def test_template_badinput(tmp_path):
+    """Testing Template() with non-conforming definitions."""
+    elements = [{}]
+    with raises(KeyError):
+        tmpl = Template(elements=elements)
+    elements = [{"name": "n", "type": "X"}]
+    with raises(KeyError):
+        tmpl = Template(elements=elements)
+        tmpl.render()
+    elements = [{"name": "n", "type": "T", "x1": 0, "y1": 0, "x2": 0, "y2": "x"}]
+    with raises(TypeError):
+        tmpl = Template(elements=elements)
+        tmpl.render()
+    tmpl = Template()
+    with raises(FPDFException):
+        tmpl.parse_csv(HERE / "mandmissing.csv", delimiter=";")
+    with raises(ValueError):
+        tmpl.parse_csv(HERE / "badint.csv", delimiter=";")
+    with raises(ValueError):
+        tmpl.parse_csv(HERE / "badfloat.csv", delimiter=";")
+    with raises(KeyError):
+        tmpl.parse_csv(HERE / "badtype.csv", delimiter=";")
+        tmpl.render()
+
+
 def test_template_code39(tmp_path):  # issue-161
     elements = [
         {
             "name": "code39",
             "type": "C39",
-            "x": 40,
-            "y": 50,
-            "h": 20,
+            "x1": 40,
+            "y1": 50,
```

Comment:
Could you please add another entry in that dict with a `code39` where neither `y1` nor `y2` ared defined?

Reply:
Without `y2` it will fail, because it can't determine its height.

Reply:
Sorry, I meant "with y1 == y2"